chain/deals/client_states.go

```diff
@@ -1,142 +1,224 @@
-		return xerrors.Errorf("deal publish message called incorrect method (method=%s)", pubmsg.Method)
+		return nil, xerrors.Errorf("deal publish message called incorrect method (method=%s)", pubmsg.Method)
+	}
+
+	var params actors.PublishStorageDealsParams
+	if err := params.UnmarshalCBOR(bytes.NewReader(pubmsg.Params)); err != nil {
+		return nil, err
+	}
+
+	dealIdx := -1
+	for i, storageDeal := range params.Deals {
+		// TODO: make it less hacky
+		eq, err := cborutil.Equals(&deal.Proposal, &storageDeal.Proposal)
+		if err != nil {
+			return nil, err
+		}
+		if eq {
+			dealIdx = i
+		}
+	}
+
+	if dealIdx == -1 {
+		return nil, xerrors.Errorf("deal publish didn't contain our deal (message cid: %s)", deal.PublishMessage)
 	}
 
 	// TODO: timeout
-	_, ret, err := c.sm.WaitForMessage(ctx, *resp.PublishMessage)
+	_, ret, err := c.sm.WaitForMessage(ctx, *deal.PublishMessage)
 	if err != nil {
-		return xerrors.Errorf("waiting for deal publish message: %w", err)
+		return nil, xerrors.Errorf("waiting for deal publish message: %w", err)
 	}
 	if ret.ExitCode != 0 {
-		return xerrors.Errorf("deal publish failed: exit=%d", ret.ExitCode)
+		return nil, xerrors.Errorf("deal publish failed: exit=%d", ret.ExitCode)
 	}
-	// TODO: persist dealId
 
-	log.Info("DEAL ACCEPTED!")
+	var res actors.PublishStorageDealResponse
+	if err := res.UnmarshalCBOR(bytes.NewReader(ret.Return)); err != nil {
+		return nil, err
+	}
 
-	return nil
+	return func(info *ClientDeal) {
+		info.DealID = res.DealIDs[dealIdx]
+	}, nil
 }
 
-func (c *Client) accepted(ctx context.Context, deal ClientDeal) error {
-	/* data transfer happens */
+func (c *Client) staged(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
+	// TODO: Maybe wait for pre-commit
 
-	resp, err := c.readStorageDealResp(deal)
-	if err != nil {
-		return err
-	}
+	return nil, nil
+}
 
-	if resp.State != api.DealStaged {
-		return xerrors.Errorf("deal wasn't staged (State=%d)", resp.State)
-	}
+func (c *Client) sealing(ctx context.Context, deal ClientDeal) (func(*ClientDeal), error) {
+	checkFunc := func(ts *types.TipSet) (done bool, more bool, err error) {
+		sd, err := stmgr.GetStorageDeal(ctx, c.sm, deal.DealID, ts)
+		if err != nil {
+			// TODO: This may be fine for some errors
+			return false, false, xerrors.Errorf("failed to look up deal on chain: %w", err)
+		}
 
-	log.Info("DEAL STAGED!")
+		if sd.ActivationEpoch > 0 {
+			select {
+			case c.updated <- clientDealUpdate{
+				newState: api.DealComplete,
+				id:       deal.ProposalCid,
+			}:
+			case <-c.stop:
+			}
 
-	return nil
-}
+			return true, false, nil
+		}
 
-func (c *Client) staged(ctx context.Context, deal ClientDeal) error {
-	/* miner seals our data, hopefully */
+		return false, true, nil
+	}
+
+	called := func(msg *types.Message, ts *types.TipSet, curH uint64) (more bool, err error) {
+		defer func() {
+			if err != nil {
+				select {
+				case c.updated <- clientDealUpdate{
```

Comment:
this feels a little bit deadlocky to me, probably fine if theres a sufficient buffer, but still nervous

Reply:
Oh, I guess these calls don't actually happen inside the client state machine loop, they happen over in the events system loop. Should be fine?

Reply:
That's a separate goroutine - https://github.com/filecoin-project/lotus/pull/472/files#diff-e264760819accf7406fce3d1ec61e7e1L16